### ISSUE TITLE
Press a key single time will arise OnKeyDown event two times

### ISSUE THREAD

**Issue by dynamicboy** [VBAndCs/sVB-Small-Visual-Basic#41]
Verrsion: 2.8.7.5
 Details:
Step1: In Form Code, input some code as below:
```Small Visual Basic
Me.OnKeyDown = Form1_OnKeyDown
Sub Form1_OnKeyDown
   TW.WriteLine("Time: " + Clock.ElapsedMilliseconds + ", You pressed " + Event.LastKey)
EndSub
```

Step2: Click Run
Step3: Press Key 0 in Numpad just one time

Output: 
![2024-01-01_003804](https://github.com/VBAndCs/sVB-Small-Visual-Basic/assets/7421525/4e951b3f-729e-4953-85b9-0491d3015e93)

**Comment by dynamicboy** [comment]
I found this behavior when I use space key to pause and resume a avatar.

**Comment by VBAndCs** [comment]
Hi @dynamicboy


You are the cause of these issue, because you insist to use sVB as if it is SB.

I told you in a previous reply that sVB uses the naming convention ControlName_EventName to handle events as VB6 does, so you can just use:

```

Sub Form1_OnKeyDown

   TW.WriteLine("Time: " + Clock.ElapsedMilliseconds + ", You pressed " + Event.LastKey)

EndSub

```


and this will handle the Form1.OnKeyDown event. I also told you that you can generate this handler sub by choosing the Form1 and OnKeyDown from the upper dropdown lists.

Now, when you add this unnecessary code:
Me.OnKeyDown = Form1_OnKeyDown

You actually register the handler twice, so it will be called two times!

And if you add the same line of code more times, you will have as many handlers as what you asked for. In fact this is the way VB .NET and C# works.

I can change this behavior, but it will need a lot of code to store a dictionary of handlers for all controls and events, so I can check if the handler is added before not to add the same handler twice, but this will need a lot of unnecessary work.

So, my advice here is to avoid registering the controls event handlers manually and let sVB do the work for you. There are two cases where you need to explicitly add the handler by code:

1. When using the SB timer or the sVB timer control.

2. when you want to use the same handler for more than one control, instead of repeating the same code for each control. For example, in the "Simple calculator" application in the sVB samples folder, the TxtNum1_OnTextInput sub is used to handle the TxtNum1.OnTextInput event, and to make txtNum2 use it also we use this code as the global area:
`txtNum2.OnTextInput = TxtNum1_OnTextInput`


I made some improvements in the sVB events system after your last issue, like fixing the Form.OnClosed event which was not fired when there is only one opened form because the program closes immediately. sVB 2.8.8 will have these fixes beside the new properties window.


You are helping a lot, so keep up the good work.

Thanks


________________________________

**Comment by dynamicboy** [comment]
@VBAndCs 
I got it, thanks for the information about event handling.
I'm trying to create as many applications as possible to learn small visual basic, based on my previous programming knowledge.
I didn't thoroughly read the reference book, and didn't notice the event droplist on the top of the code form, my bad.

I just corrected my code, the pause and resume mechanism is working now.

**Comment by VBAndCs** [comment]
I decided to fix this issue, because sVB syntax deon't allow to remove event handlers (while VB .NET does). I am working on it for more than 3 hours now, but it will allow sVB programmers to change the event handler at anytime while the program is running without suffering any side effects. All these fixes will be valid soon in sVB 2.8.8


________________________________
